### DISCUSSION TITLE
socket.io: please allow @loopback/socket.io to work with rest applications

### DISCUSSION THREAD

**Discussion by clewisln** [loopbackio/loopback-next#8267] [Ideas]
Socket.io is a great tool for push notifications. However, the current implementation of @loopback/socketio assumes it will be responsible for creating the http server itself.

Socket.io supports attaching to an _existing_ http server, including the one created by RestApplication. These two features should be allowed to work together, to deliver both a restful api, and socket-based event subscriptions, from the same ip and port.

@loopback/socketio needs to be refactored to allow an external server class to be configured, such as "RestServer". The idea is that the server instance is retrieved from the current context, when the _start_ observer is fired, which socket.io is then attached to. If this configuration is not provided, then falling back to creating its own http server is fine. 

Example implementation of the observer (simplified and incomplete):
```
async start() {
    //Retrieve the http server instance created by the rest server, then attach socket.io onto it.
    const appServer = await this.app.getServer(RestServer);
    if (!appServer.httpServer)
      throw new Error("SocketIO: Application http server is not available to attach to.");
    this.io.attach(appServer.httpServer.server, this.options.socketIoOptions);
}
```

Also: SocketIoComponent will have to register SocketIoServer as a life cycle observer, instead of as a server, to prevent RestApplication from throwing an exception, when using attach mode.

**Comment by mpyrkosz88** [comment]
Hello,
Any updates how to connect socket.io with rest application? Some examples?

**Reply by clewisln** [reply]
I am unable to provide a full example of my solution, as the code is confidential.

However, the original post outlined the high level flow of the implementation my project uses.

I.e. taking a copy of the socket io server class provided by loopback, as a base, modifying it into a lifecycle observer, then grabbing the server instance and manually attaching to it. The down side is that I was not able to figure out a proper way to implement sticky sessions using the loopback workflow, so only websockets are supported in a multi-server load-balanced environment.

**Reply by mpyrkosz88** [reply]
Thank for answer. I tried to use on this way. I was thinking to use older version of socketIO, from this example.
https://github.com/raymondfeng/loopback4-example-websocket/
Have you tried it?
Maybe you could try to create new clean repo and just copy code from socketIO ?
